File: internal/skillscanner/skillscanner_test.go
package skillscanner

import (
	"os"
	"path/filepath"
	"strings"
	"testing"
)

func writeSkill(t *testing.T, root, name, frontmatter, body string) string {
	t.Helper()
	dir := filepath.Join(root, name)
	if err := os.MkdirAll(dir, 0o755); err != nil {
		t.Fatalf("mkdir: %v", err)
	}
	content := "---\n" + frontmatter + "\n---\n" + body
	if err := os.WriteFile(filepath.Join(dir, "SKILL.md"), []byte(content), 0o644); err != nil {
		t.Fatalf("write: %v", err)
	}
	return dir
}

func findSkill(skills []skillLike, name string) *skillLike {
	for i := range skills {
		if skills[i].Name == name {
			return &skills[i]
		}
	}
	return nil
}

// skillLike lets us share a find helper without importing the models package
// twice via a different name.
type skillLike struct {
	Name        string
	Description string
	Path        string
	Valid       bool
	Errors      []string
	Frontmatter map[string]string
	BodyPreview string
}

func scanAs(t *testing.T, root string) []skillLike {
	t.Helper()
	skills, err := ScanSkills(root)
	if err != nil {
		t.Fatalf("ScanSkills: %v", err)
	}
	out := make([]skillLike, len(skills))
	for i, s := range skills {
		out[i] = skillLike{
			Name:        s.Name,
			Description: s.Description,
			Path:        s.Path,
			Valid:       s.Valid,
			Errors:      s.Errors,
			Frontmatter: s.Frontmatter,
			BodyPreview: s.BodyPreview,
		}
	}
	return out
}

func TestScanSkillsValid(t *testing.T) {
	root := t.TempDir()
	writeSkill(t, root, "valid-skill", "name: valid-skill\ndescription: A nice skill for testing.", "\n# Valid Skill\n\nHello.")
	skills := scanAs(t, root)
	found := findSkill(skills, "valid-skill")
	if found == nil {
		t.Fatal("valid-skill not found")
	}
	if !found.Valid {
		t.Errorf("expected valid=true, got errors: %v", found.Errors)
	}
	if found.Description != "A nice skill for testing." {
		t.Errorf("unexpected description: %q", found.Description)
	}
}

func TestScanSkillsMissingDescription(t *testing.T) {
	root := t.TempDir()
	writeSkill(t, root, "missing-desc", "name: missing-desc", "\n# Body only\n")
	skills := scanAs(t, root)
	found := findSkill(skills, "missing-desc")
	if found == nil {
		t.Fatal("missing-desc not found")
	}
	if found.Valid {
		t.Errorf("expected valid=false")
	}
	hasDescErr := false
	for _, e := range found.Errors {
		if strings.Contains(strings.ToLower(e), "description") {
			hasDescErr = true
			break
		}
	}
	if !hasDescErr {
		t.Errorf("expected an error mentioning description, got %v", found.Errors)
	}
}

func TestScanSkillsIgnoresNonSkill(t *testing.T) {
	root := t.TempDir()
	folder := filepath.Join(root, "not-a-skill")
	if err := os.MkdirAll(folder, 0o755); err != nil {
		t.Fatal(err)
	}
	if err := os.WriteFile(filepath.Join(folder, "readme.txt"), []byte("hello"), 0o644); err != nil {
		t.Fatal(err)
	}
	skills := scanAs(t, root)
	for _, s := range skills {
		if s.Path == folder {
			t.Fatal("non-skill folder was scanned as a skill")
		}
	}
}

func TestScanSkillsBodyPreviewTruncated(t *testing.T) {
	root := t.TempDir()
	longBody := "\n" + strings.Repeat("a", 1500)
	writeSkill(t, root, "long-body", "name: long-body\ndescription: A long body skill.", longBody)
	skills := scanAs(t, root)
	found := findSkill(skills, "long-body")
	if found == nil {
		t.Fatal("long-body not found")
	}
	if len(found.BodyPreview) != BodyPreviewLimit {
		t.Errorf("expected bodyPreview length %d, got %d", BodyPreviewLimit, len(found.BodyPreview))
	}
}

func TestParseSkillMarkdownMissingFile(t *testing.T) {
	root := t.TempDir()
	parsed, err := ParseSkillMarkdown(filepath.Join(root, "does-not-exist.md"))
	if err != nil {
		t.Fatalf("unexpected error: %v", err)
	}
	if len(parsed.Errors) == 0 {
		t.Fatal("expected errors for missing file")
	}
	if parsed.Name != "" {
		t.Errorf("expected empty name, got %q", parsed.Name)
	}
}

func TestImportSkillFolderNoOverwrite(t *testing.T) {
	root := t.TempDir()
	source := t.TempDir()
	if err := os.WriteFile(
		filepath.Join(source, "SKILL.md"),
		[]byte("---\nname: ported\ndescription: Imported skill.\n---\nBody"),
		0o644,
	); err != nil {
		t.Fatal(err)
	}

	if _, err := ImportSkillFolder(source, root); err != nil {
		t.Fatalf("first import failed: %v", err)
	}
	if _, err := ImportSkillFolder(source, root); err == nil {
		t.Fatal("second import should have failed (target exists)")
	}
}

File: tests/skillScanner.test.ts
import { describe, it, expect, beforeAll, afterAll } from "vitest";
import { promises as fs } from "node:fs";
import os from "node:os";
import path from "node:path";
import { parseSkillMarkdown, scanSkills } from "../src/main/skillScanner";

let tmpRoot: string;

beforeAll(async () => {
  tmpRoot = await fs.mkdtemp(path.join(os.tmpdir(), "skill-scanner-"));
});

afterAll(async () => {
  await fs.rm(tmpRoot, { recursive: true, force: true });
});

async function writeSkill(name: string, frontmatter: string, body: string): Promise<string> {
  const dir = path.join(tmpRoot, name);
  await fs.mkdir(dir, { recursive: true });
  const content = `---\n${frontmatter}\n---\n${body}`;
  await fs.writeFile(path.join(dir, "SKILL.md"), content, "utf8");
  return dir;
}

describe("skillScanner", () => {
  it("parses a valid SKILL.md and marks it valid", async () => {
    await writeSkill(
      "valid-skill",
      "name: valid-skill\ndescription: A nice skill for testing.",
      "\n# Valid Skill\n\nHello.",
    );
    const skills = await scanSkills(tmpRoot);
    const found = skills.find((s) => s.name === "valid-skill");
    expect(found).toBeTruthy();
    expect(found!.valid).toBe(true);
    expect(found!.description).toBe("A nice skill for testing.");
  });

  it("marks a skill invalid when description is missing", async () => {
    await writeSkill(
      "missing-desc",
      "name: missing-desc",
      "\n# Body only\n",
    );
    const skills = await scanSkills(tmpRoot);
    const found = skills.find((s) => s.name === "missing-desc");
    expect(found).toBeTruthy();
    expect(found!.valid).toBe(false);
    expect(found!.errors.some((e) => e.toLowerCase().includes("description"))).toBe(true);
  });

  it("ignores folders that do not contain SKILL.md", async () => {
    const folder = path.join(tmpRoot, "not-a-skill");
    await fs.mkdir(folder, { recursive: true });
    await fs.writeFile(path.join(folder, "readme.txt"), "hello", "utf8");
    const skills = await scanSkills(tmpRoot);
    expect(skills.find((s) => s.path === folder)).toBeUndefined();
  });

  it("truncates body preview to 800 characters", async () => {
    const longBody = "\n" + "a".repeat(1500);
    await writeSkill(
      "long-body",
      "name: long-body\ndescription: A long body skill.",
      longBody,
    );
    const skills = await scanSkills(tmpRoot);
    const found = skills.find((s) => s.name === "long-body");
    expect(found).toBeTruthy();
    expect(found!.bodyPreview.length).toBe(800);
  });

  it("parseSkillMarkdown returns errors but no throw for missing file", async () => {
    const parsed = await parseSkillMarkdown(path.join(tmpRoot, "does-not-exist.md"));
    expect(parsed.errors.length).toBeGreaterThan(0);
    expect(parsed.name).toBe("");
  });
});
